File: sprintengine_mcp/protocol.py
"""The MCP protocol versions this engine can honestly serve.

One table, imported by every transport (`server.py` stdio dispatch,
`http_server.py`), because the previous inline constants are exactly how the two
Multicode MCP servers drifted apart — the socket gateway defaulted to
`2025-03-26` while this engine defaulted to `2024-11-05`.

A version belongs here only once the semantics behind it exist: answering
`initialize` with a version we do not implement is the lie this module was
created to remove. The TypeScript half is `src/shared/mcp/protocol.ts` and
carries the same set.
"""

from __future__ import annotations

from typing import cast

# Newest-first. The head is what we answer when the client's ask is not on the list.
# `2026-07-28` is earned by the HTTP transport, not assumed: session-less requests
# (SEP-2575/SEP-2567), the per-request version declaration, the `Mcp-Method`/`Mcp-Name`
# headers (SEP-2243), and `ttlMs` on `tools/list` (SEP-2549) all land in `http_server.py`
# and `server.py` in the same commit that prepends it here.
SUPPORTED_PROTOCOL_VERSIONS: tuple[str, ...] = ("2026-07-28", "2025-06-18", "2025-03-26", "2024-11-05")

DEFAULT_PROTOCOL_VERSION: str = SUPPORTED_PROTOCOL_VERSIONS[0]

# What the spec says an HTTP request means when it declares no version at all.
LEGACY_HTTP_ASSUMED_VERSION = "2025-03-26"


def is_supported_protocol_version(value: object) -> bool:
    return isinstance(value, str) and value in SUPPORTED_PROTOCOL_VERSIONS


def negotiate_protocol_version(requested: object) -> str:
    """Answer a supported request with itself, anything else with the default.

    Never raises. The spec's rule for an unsupported `initialize` is to respond
    with a version the server does support and let the client decide whether to
    continue; an error here would break clients that would have accepted our
    answer. `requested` is deliberately `object`: it arrives straight off the
    wire, so a missing key, `None`, or a non-string all land in the same branch.
    """
    if is_supported_protocol_version(requested):
        return cast(str, requested)
    return DEFAULT_PROTOCOL_VERSION

File: src/shared/mcp/protocol.ts
// The MCP protocol versions the TypeScript side can honestly serve.
//
// One table, imported by the socket gateway (`src/main/automation/mcp-socket-server.ts`)
// and by every other MCP speaker in the app, because inline constants are
// exactly how two MCP servers in one product drift apart — one fell back to
// `2025-03-26` while the other defaulted to `2024-11-05`.
//
// A version belongs here only once the semantics behind it exist: answering
// `initialize` with a version we do not implement is the lie this module was
// created to remove. Lives in `shared` — no node-only imports — beside
// `normalize-server.ts`, because a main-process server and a main-process
// client both need it.

/** Newest-first. The head is what we answer when the client's ask is not on the list. */
export const SUPPORTED_MCP_PROTOCOL_VERSIONS: readonly string[] = [
  // Earned by the socket gateway (item 2141): handshake-optional framing, the
  // per-request `_meta.protocolVersion` declaration, and `ttlMs` on tools/list.
  '2026-07-28',
  '2025-06-18',
  '2025-03-26',
  '2024-11-05',
]

export const DEFAULT_MCP_PROTOCOL_VERSION: string = SUPPORTED_MCP_PROTOCOL_VERSIONS[0]

export function isSupportedMcpProtocolVersion(value: unknown): value is string {
  return typeof value === 'string' && SUPPORTED_MCP_PROTOCOL_VERSIONS.includes(value)
}

/**
 * Answer a supported request with itself, an unknown one with the newest
 * version we serve that is no NEWER than the ask.
 *
 * Never throws. The spec's rule for an unsupported `initialize` is to respond
 * with a version the server does support and let the client decide whether to
 * continue; an error here would break clients that would have accepted our
 * answer. `requested` is `unknown` because it arrives straight off the wire, so
 * an absent field, `null`, or a non-string all land in the same branch.
 *
 * Answering the unknown ask with our MAXIMUM is what broke Claude Code: it
 * offers a revision between `2025-06-18` and `2026-07-28`, so handing back
 * `2026-07-28` names a version newer than the client can parse and the
 * handshake is rejected outright — every agent we spawned lost its tools. Downgrading instead lands on `2025-06-18`, which both sides speak.
 * Versions are ISO dates, so lexicographic order is chronological order.
 */
export function negotiateMcpProtocolVersion(requested: unknown): string {
  if (isSupportedMcpProtocolVersion(requested)) return requested
  if (typeof requested !== 'string' || requested === '') return DEFAULT_MCP_PROTOCOL_VERSION
  // Newest-first, so the first entry at or below the ask is the best downgrade.
  const downgrade = SUPPORTED_MCP_PROTOCOL_VERSIONS.find((version) => version <= requested)
  // No entry at or below means the client is older than anything we serve;
  // answer our oldest and let it decide, rather than our newest.
  return downgrade ?? SUPPORTED_MCP_PROTOCOL_VERSIONS[SUPPORTED_MCP_PROTOCOL_VERSIONS.length - 1]
}
